doubly_linked_list/doubly_linked_list.py

```diff
@@ -40,25 +40,70 @@ def __init__(self, node=None):
     self.tail = node
 
   def add_to_head(self, value):
-    pass
+    new_node = ListNode(value)
+    if self.head == None:
+      self.head = new_node
+    else:
+      current_head = self.head.value
+      self.head = new_node
+      self.head.insert_after(current_head) # moving the current head to next node
+
+    
 
   def remove_from_head(self):
-    pass
+    current_head = self.head
+    if self.head == None:
+      return None
+    else:
+      self.head.delete()
+      return current_head.value
 
   def add_to_tail(self, value):
-    pass
+    new_node = ListNode(value)
+    if self.tail == None:
+      self.tail = new_node
+    else:
+      prev_tail = self.tail.value
+      self.tail = new_node
+      self.tail.prev = prev_tail
+      self.tail.insert_before(self.tail.value)
```

Comment:
It looks like you're on the right track here. You seem to have a problem passing the tests because of this method. Here's what I would do: If there's no tail, set the tail to a ListNode with the value passed in and the head as the prev. `elif` there's `not` a head, set the head to the tail, set the tail to a ListNode with the value passed in and the head as the prev, then set the head's `next` to the tail. `else` set the tail to a ListNode with the value passed in and the tail as the prev, then set the tail's prev next to the tail.